packages/vulcan-forms/lib/components/Form.jsx

```diff
@@ -450,22 +477,93 @@ class Form extends Component {
       return newState;
     });
   };
-
+  
+  /*
+  
+  Warn the user if there are unsaved changes
+  
+  */
+  handleRouteLeave = () => {
+    if (this.isChanged()) {
+      const message = this.context.intl.formatMessage({
+        id: 'forms.confirm_discard',
+        defaultMessage: 'Are you sure you want to discard your changes?'
+      });
+      return message;
+    }
+  };
+  
+  /*
+  
+  Install a route leave hook to warn the user if there are unsaved changes
+  
+  */
+  componentDidMount = () => {
+    let warnUnsavedChanges = getSetting('forms.warnUnsavedChanges');
+    if (typeof this.props.warnUnsavedChanges === 'boolean') {
+      warnUnsavedChanges = this.props.warnUnsavedChanges;
+    }
+    if (warnUnsavedChanges) {
+      const routes = this.props.router.routes;
+      const currentRoute = routes[routes.length - 1];
+      this.props.router.setRouteLeaveHook(currentRoute, this.handleRouteLeave);
+    }
+  };
+  
   /*
+  
+  Returns true if there are any differences between the initial document and the current one
+  
+  */
+  isChanged = () => {
+    const initialDocument = this.state.initialDocument;
+    const changedDocument = this.getDocument();
+    
+    const changedValue = find(changedDocument, (value, key, collection) => {
+      return !isEqualWith(value, initialDocument[key], (objValue, othValue) => {
+        if (!objValue && !othValue) return true;
+      });
+    });
+    
+    return typeof changedValue !== 'undefined';
+  };
```

Comment:
I think maybe I asked you before, but why not just look at `currentValues` and `deletedValues` to see if there's any changes?

Reply:
1. If there is anything in defaultValues, isChanged would always be true (even if nothing changed)

2. Entering a value and then clearing it or clearing a value and then entering the same value does not leave currentValues and deletedValues in the same state.